homeassistant/components/media_player/mediaroom.py

```diff
@@ -9,134 +9,169 @@
+
+    def callback_notify(notify):
+        """Process NOTIFY message from STB."""
+        if notify.ip_address in hass.data[DATA_MEDIAROOM]:
+            dispatcher_send(hass, SIGNAL_STB_NOTIFY, notify)
             return
-    hosts.append(host)
-    known_hosts.append(host)
-
-    stbs = []
 
-    try:
-        for host in hosts:
-            stbs.append(MediaroomDevice(
-                config.get(CONF_NAME),
-                host,
-                config.get(CONF_OPTIMISTIC),
-                config.get(CONF_TIMEOUT)
-            ))
+        _LOGGER.debug("Discovered new stb %s", notify.ip_address)
+        hass.data[DATA_MEDIAROOM].append(notify.ip_address)
+        new_stb = MediaroomDevice(
+            host=notify.ip_address, device_id=notify.device_uuid,
+            optimistic=False
+        )
+        async_add_devices([new_stb])
 
-    except ConnectionRefusedError:
-        hass.components.persistent_notification.create(
-            'Error: Unable to initialize mediaroom at {}<br />'
-            'Check its network connection or consider '
-            'using auto discovery.<br />'
-            'You will need to restart hass after fixing.'
-            ''.format(host),
-            title=NOTIFICATION_TITLE,
-            notification_id=NOTIFICATION_ID)
+    if not config.get(CONF_OPTIMISTIC):
+        from pymediaroom import install_mediaroom_protocol
 
-    add_devices(stbs)
+        await install_mediaroom_protocol(responses_callback=callback_notify)
+        _LOGGER.debug("Auto discovery installed")
 
 
 class MediaroomDevice(MediaPlayerDevice):
     """Representation of a Mediaroom set-up-box on the network."""
 
-    def __init__(self, name, host, optimistic=False, timeout=DEFAULT_TIMEOUT):
+    from pymediaroom import State
+
+    STATE_MAP = {
+        State.OFF: STATE_OFF,
+        State.STANDBY: STATE_STANDBY,
+        State.PLAYING_LIVE_TV: STATE_PLAYING,
+        State.PLAYING_RECORDED_TV: STATE_PLAYING,
+        State.PLAYING_TIMESHIFT_TV: STATE_PLAYING,
+        State.STOPPED: STATE_PAUSED,
+        State.UNKNOWN: STATE_UNAVAILABLE
+        }
+
+    def __init__(self, host, device_id, optimistic=False,
+                 timeout=DEFAULT_TIMEOUT):
         """Initialize the device."""
         from pymediaroom import Remote
 
-        self.stb = Remote(host, timeout=timeout)
-        _LOGGER.info(
-            "Found %s at %s%s", name, host,
-            " - I'm optimistic" if optimistic else "")
-        self._name = name
-        self._is_standby = not optimistic
-        self._current = None
+        self.host = host
+        self.stb = Remote(host)
+        _LOGGER.info("Found STB at %s%s", host,
+                     " - I'm optimistic" if optimistic else "")
+        self._channel = None
         self._optimistic = optimistic
-        self._state = STATE_STANDBY
+        self._state = STATE_PLAYING if optimistic else STATE_STANDBY
+        self._name = '{}_{}'.format(DOMAIN, device_id)
```

Comment:
1. See other comment about `DOMAIN`.
2. I don't think friendly names should have an underscore in them.